### ISSUE TITLE
IndexCompilationException when using OrderBy().ThenBy()

### ISSUE THREAD

**Issue by joaope** [ravendb/ravendb#12431]
Trying to use `ThenBy()` together with `OrderBy()` always results in a `IndexCompilationException` as such:

> error CS1929: 'IEnumerable<dynamic>' does not contain a definition for 'ThenBy' and the best extension method overload 'Enumerable.ThenBy<dynamic, dynamic>(IOrderedEnumerable<dynamic>, Func<dynamic, dynamic>)' requires a receiver of type 'IOrderedEnumerable<dynamic>'
    , IndexDefinitionProperty='', ProblematicText=''   at Raven.Server.Documents.Indexes.Static.IndexCompiler.CompileInternal(String originalName, String cSharpSafeName, MemberDeclarationSyntax class, Dictionary`2 extensions) in C:\Builds\RavenDB-Stable-4.2\42062\src\Raven.Server\Documents\Indexes\Static\IndexCompiler.cs:line 197
       at Raven.Server.Documents.Indexes.Static.IndexCompiler.Compile(IndexDefinition definition) in C:\Builds\RavenDB-Stable-4.2\42062\src\Raven.Server\Documents\Indexes\Static\IndexCompiler.cs:line 121
       at Raven.Server.Documents.Indexes.Static.IndexCompilationCache.GenerateIndex(IndexDefinition definition, RavenConfiguration configuration, IndexType type) in C:\Builds\RavenDB-Stable-4.2\42062\src\Raven.Server\Documents\Indexes\Static\IndexCompilationCache.cs ...

Snippet that results in an expection:

```c#
public class TestIndexWithThenBy : AbstractIndexCreationTask<Document>
{
    public TestIndexWithThenBy()
    {
        Map = docs => docs.Select(doc => new { Prop1 = doc.Prop1, Prop2 = doc.Prop2, Prop3 = doc.Prop3 });
        Reduce = results => results
            .GroupBy(result => new { result.Prop1 })
            .Select(result => result.OrderBy(x => x.Prop2).ThenBy(x => x.Prop3).First())
            .Select(result => new { Prop1 = result.Prop1, Prop2 = result.Prop2, Prop3 = result.Prop3 });
    }
}
```

It works if I instead use `OrderBy().OrderBy()` but that is obviously semantically incorrect and will most likely impact performance.

Looks like #11962 tries to fix it but it doesn't. Also, [DynamicEnumerable](https://github.com/ppekrol/ravendb/blob/c973e25fa5f3dd6bb0183f0f51f8dcdee99b02e1/src/Raven.Server/Documents/Indexes/Static/Linq/DynamicEnumerable.cs#L182-L190) doesn't have any `ThenBy` extension method but I'm not sure if that's intentional.

Currently using the very latest of `RavenDB.Client` **4.2.114** and going backwards it only starts working again with _4.2.101_.

Thank you!

**Comment by ppekrol** [comment]
This PR https://github.com/ravendb/ravendb/pull/11962 is a server-side fix that was applied to RavenDB v5.1. Is there a possibility for you to use v5.1 (or newer) server?

**Comment by joaope** [comment]
> This PR #11962 is a server-side fix that was applied to RavenDB v5.1. Is there a possibility for you to use v5.1 (or newer) server?

Hi. Thanks for the quick reply.

Upgrading is not really something is in our roadmap for the near future. Licensing and all is a bit out of my control also tbh to have a clear idea if that's happening.

Any way we can backport that fix to 4.x? We haven't been able to update the client's packages for over a year now.

**Comment by ppekrol** [comment]
Here you go: https://github.com/ravendb/ravendb/pull/12434

**Comment by joaope** [comment]
> Here you go: #12434

This is brilliant! Really appreciate the quick response.

**Comment by ppekrol** [comment]
After this is merged I can create a nightly build for you to test it out. What do you think? The fix itself will be available in next stable (planned for end of next week).

**Comment by joaope** [comment]
> After this is merged I can create a nightly build for you to test it out. What do you think? The fix itself will be available in next stable (planned for end of next week).

Absolutely. Let me know from where I can get the nightly from and I can give it a go.

**Comment by ppekrol** [comment]
We have released `4.2.115-nightly-20210625-0951` you can download it from https://hibernatingrhinos.com/builds?product=RavenDB&version=4.2&channel=Nightly&pageNr=1

**Comment by joaope** [comment]
> We have released `4.2.115-nightly-20210625-0951` you can download it from https://hibernatingrhinos.com/builds?product=RavenDB&version=4.2&channel=Nightly&pageNr=1

Just tested this out and it fixes the issue!

I forgot to mention initially that we are also using `RavenDB.TestDriver` which is actually where we see the issue manifesting itself first, during the tests' stage. I've also tested it out by replacing the embedded server in `/RavenDBServer`.

I assume your fix will also make it to the embedded server through the `RavenDB.Embedded` package.

Thank you again for the quick fix.

**Comment by ppekrol** [comment]
Glad that it helped. You are correct, `RavenDB.Embedded` will contain that fix as well.
